janus_server/src/datastore.rs

```diff
@@ -615,6 +615,63 @@ impl<C: Clock> Transaction<'_, C> {
             .collect::<Result<Vec<Nonce>, Error>>()
     }
 
+    /// get_unaggregated_client_report_nonces_by_collect_for_task returns pairs of nonces and
+    /// aggregation parameters, corresponding to client reports that have not yet been aggregated,
+    /// or not aggregated with a certain aggregation parameter, and for which there are collect
+    /// jobs, for a given task.
+    #[cfg(test)]
+    #[tracing::instrument(skip(self), err)]
+    pub async fn get_unaggregated_client_report_nonces_by_collect_for_task<
+        const L: usize,
+        A: vdaf::Aggregator<L>,
+    >(
+        &self,
+        task_id: TaskId,
+    ) -> Result<Vec<(Nonce, A::AggregationParam)>, Error>
+    where
+        for<'a> &'a A::AggregateShare: Into<Vec<u8>>,
+    {
+        let stmt = self
+            .tx
+            .prepare_cached(
+                "SELECT DISTINCT nonce_time, nonce_rand, collect_jobs.aggregation_param
+                FROM collect_jobs
+                INNER JOIN client_reports
+                ON collect_jobs.task_id = client_reports.task_id
+                AND client_reports.nonce_time <@ collect_jobs.batch_interval
+                LEFT JOIN (
+                    SELECT report_aggregations.id, report_aggregations.client_report_id,
+                        aggregation_jobs.aggregation_param
+                    FROM report_aggregations
+                    INNER JOIN aggregation_jobs
+                    ON aggregation_jobs.id = report_aggregations.aggregation_job_id
+                    WHERE aggregation_jobs.task_id = (SELECT id FROM tasks WHERE task_id = $1)
+                ) AS report_aggs
+                ON report_aggs.client_report_id = client_reports.id
+                AND report_aggs.aggregation_param = collect_jobs.aggregation_param
+                WHERE client_reports.task_id = (SELECT id FROM tasks WHERE task_id = $1)
+                AND collect_jobs.task_id = (SELECT id FROM tasks WHERE task_id = $1)
+                AND collect_jobs.state = 'START'
+                AND report_aggs.id IS NULL
+                ORDER BY nonce_time DESC LIMIT 5000",
```

Comment:
`LIMIT 5000` isn't going to work forever. We have a similar clause on the query `get_unaggregated_client_report_nonces_for_task` that references #224 though I admit I don't quite see how this relates to performance and load testing (I guess it's under load that looking at only the last 5,000 results would fail). Anyway let's make sure we've made a note somewhere that this needs to be revisited.